connector.go

```diff
@@ -0,0 +1,264 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/UniversityRadioYork/baps3-go"
+)
+
+// GetResponse is the outer structure of all GET responses.
+type GetResponse struct {
+	Status string
+	Value  interface{}
+}
+
+// GetOk creates a GetResponse wrapping a successful GET result.
+func GetOk(value interface{}) *GetResponse {
+	r := new(GetResponse)
+	r.Status = "ok"
+	r.Value = value
+	return r
+}
+
+type bfConnector struct {
+	conn   *baps3.Connector
+	name   string
+	wg     *sync.WaitGroup
+	logger *log.Logger
+	state  *serviceState
+
+	reqCh chan httpRequest
+	resCh <-chan baps3.Message
+
+	// TODO(CaptainHayashi): move this away from baps3.Message to
+	// something generic.
+	updateCh chan<- baps3.Message
+}
+
+func initBfConnector(name string, updateCh chan baps3.Message, waitGroup *sync.WaitGroup, logger *log.Logger) (c *bfConnector) {
+	resCh := make(chan baps3.Message)
+
+	c = new(bfConnector)
+	c.resCh = resCh
+	c.conn = baps3.InitConnector(name, resCh, waitGroup, logger)
+	c.name = name
+	c.wg = waitGroup
+	c.logger = logger
+	c.reqCh = make(chan httpRequest)
+	c.updateCh = updateCh
+	c.state = initServiceState()
+	return
+}
+
+func (c *bfConnector) Run() {
+	defer c.wg.Done()
+	defer close(c.conn.ReqCh)
+
+	go c.conn.Run()
+
+	fmt.Printf("connector %s now listening for requests\n", c.name)
+
+	for {
+		select {
+		case rq, ok := <-c.reqCh:
+			if !ok {
+				return
+			}
+
+			// TODO(CaptainHayashi): probably make this more robust
+			resource := strings.Replace(rq.resource, "/"+c.name, "", 1)
+			fmt.Printf("connector %s response %s\n", c.name, resource)
+
+			// TODO(CaptainHayashi): other methods
+			rq.resCh <- c.get(resource)
+		case res := <-c.resCh:
+			if err := c.state.update(res); err != nil {
+				fmt.Println(err)
+			}
+			c.updateCh <- res
+		}
+	}
+
+	return
+}
+
+func splitResource(resource string) []string {
+	res := strings.Split(strings.Trim(resource, "/"), "/")
+
+	// The empty resource is returned as {""}: let's fix that
+	if len(res) == 1 && res[0] == "" {
+		res = []string{}
+	}
+
+	return res
+}
+
+func (c *bfConnector) get(resource string) interface{} {
+	// TODO(CaptainHayashi): HTTP status codes
+
+	resourcePath := splitResource(resource)
+
+	r := c.rootGet(resourcePath)
+
+	if r == nil {
+		// TODO(CaptainHayashi): more errors
+		return GetResponse{
+			Status: "what",
+			Value:  "resource not found: " + resource,
+		}
+	}
+
+	return GetOk(r)
+}
+
+/* These resMaps describe simple composite resources, mapping each child
+ * resource to the functions handling them.
+ *
+ * TODO(CaptainHayashi): add support for things that aren't GET
+ *   (have each resource be a jump table of possible methods, or send the method
+ *    to the resMap func?)
+ * TODO(CaptainHayashi): decouple traversal from GET
+ * TODO(CaptainHayashi): maybe make traversal iterative instead of recursive?
+ */
+
+type resMap map[string]func(*bfConnector, []string) interface{}
```

Comment:
ew. I say this, i'm probably going to use something similar for listd at some point